a08.evaluator/Evaluator.cs

```diff
@@ -38,31 +54,30 @@ class Evaluator {
 
    void Process (Token token) {
       switch (token) {
-         case TNumber num: 
-            mOperands.Push (num.Value); 
+         case TNumber num:
+            mOperands.Push (num.Value);
             break;
          case TOperator op:
-            while (mOperators.Count > 0 && mOperators.Peek ().Priority > op.Priority)
-               ApplyOperator ();
+            while (mOperators.Count != 0 && mOperands.Count != 0 && mOperators.Peek ().Priority >= op.Priority) ApplyOperator ();
             mOperators.Push (op);
             break;
-         case TPunctuation p:
-            BasePriority += p.Punct == '(' ? 10 : -10;
+         case TPunctuation p when p.Punct is '(':
             break;
-         default:
-            throw new EvalException ($"Unknown token: {token}");
+         case TPunctuation p when p.Punct is ')':
+            ApplyOperator (); break;
       }
    }
    readonly Stack<double> mOperands = new ();
    readonly Stack<TOperator> mOperators = new ();
 
    void ApplyOperator () {
       var op = mOperators.Pop ();
-      var f1 = mOperands.Pop ();
-      if (op is TOpFunction func) mOperands.Push (func.Evaluate (f1));
-      else if (op is TOpArithmetic arith) {
-         var f2 = mOperands.Pop ();
+      double f1 = mOperands.Pop (), f2;
+      if (op is TOpArithmetic arith) {
+         try {
+            f2 = mOperands.Pop ();
+         } catch (Exception) { throw new EvalException ("Too few operands"); }

```

Comment:
You are using try catch to check whether you have only one operand. Simple method is to check whether mOperands.IsEmpty and then throw the exception. Another method is to catch specific exception as below:
```suggestion
         try {
            f2 = mOperands.Pop ();
         } catch (InvalidOperationException) { throw new EvalException ("Too few operands"); }

```
In the above case if stack is empty, it throws InvalidOperationException.

Reply:
In general, we have to code in such a way exceptions are rarely thrown. More details [here](https://www.joelonsoftware.com/2003/10/13/13/).